### ISSUE TITLE
[BUG] PT_LOAD past end of file. ("-fuse-ld=lld")

### ISSUE THREAD

**Issue by SanjaLV** [android/ndk#1148]
#### Description

Building with ndk-build and `-fuse-ld=lld` generates .so file that, contains `PT_LOAD` past end of file.
Fails on NDK r21 (both betas). Does not fail with GNU linker (in that case `.bss` section is part of 01 segment with `.fini_array .data.rel.ro .dynamic .got`).
```
ls -l libs/arm64-v8a/
10424 libtest.so

readelf -l libs/arm64-v8a/libtest.so
Program Headers:
  Type           Offset             VirtAddr           PhysAddr
                 FileSiz            MemSiz              Flags  Align

  LOAD(04)       0x0000000000003000 0x0000000000003000 0x0000000000003000
                 0x0000000000000000 0x00000000003d0900  RW     0x1000

Segment Sections...

04     .bss

```
PT_LOAD offset is larger than file size.
12 288(0x3000) > 10 424.

Android Runtime will crash with error like this:
```
java.lang.UnsatisfiedLinkError: dlopen failed: invalid ELF file "/data/app/package-name/lib/arm64/libtest.so"
load segment[4]: p_offset (0x10000) + p_filesz (0x0) ( = 0x10000) past end of file (0xff00)
```

NDK r20b does not have this issue, however .so file size is **huge** (Probably due to 64K alignment).

This looks like related binutils issue:
https://sourceware.org/bugzilla/show_bug.cgi?id=25237

Minimal ndk-build repro:
https://github.com/SanjaLV/bss-load-repro

#### Environment Details
* NDK Version: r21-beta2
* Build system: ndk-build
* Host OS: Ubuntu 19.04 disco; Kernel: x86_64 Linux 5.0.0-36-generic

* ABI: Not ABI specific. (In repro for simplicity is used arm64-v8a)

**Comment by rprichard** [comment]
@MaskRay @rui314 Maybe this is an lld bug? The Bionic loader is currently rejecting a PT_LOAD segment where `p_offset + p_filesz` is larger than the file's actual size, measured in bytes. In this case, `p_offset` is the file size rounded up to the next 0x1000 page boundary, and `p_filesz` is zero.

**Comment by rprichard** [comment]
Oh wait, I think lld is fine, and it's only the (binutils) `strip` tool creating PT_LOAD segments like this.

**Comment by rprichard** [comment]
The version of lld in r21-beta2 (5900059 based on r365631c) produces a PT_LOAD segment with a 0 filesz where p_offset(0x21f0) and p_vaddr(0x3000) aren't congruent with respect to p_align(0x1000):
```
$ readelf -lW   obj/local/arm64-v8a/libtest.so
...
  Type           Offset   VirtAddr           PhysAddr           FileSiz  MemSiz   Flg Align
...
  LOAD           0x000000 0x0000000000000000 0x0000000000000000 0x000498 0x000498 R   0x1000
  LOAD           0x001000 0x0000000000001000 0x0000000000001000 0x000080 0x000080 R E 0x1000
  LOAD           0x002000 0x0000000000002000 0x0000000000002000 0x0001f0 0x0001f0 RW  0x1000
  LOAD           0x0021f0 0x0000000000003000 0x0000000000003000 0x000000 0x3d0900 RW  0x1000
```

 Using upstream lld, though, I don't see a load segment like that, and I also don't see the `p_offset + p_filesz > file_size` problem with upstream lld and binutils strip. Given the open bugzilla issue, I doubt the root `strip` problem is fixed, though.

The files generated by upstream lld are much smaller than those in r21-beta2. I think [D64906](https://reviews.llvm.org/D64906) packed the segments together more closely. For example, the load segments above become:

```
  Type           Offset   VirtAddr           PhysAddr           FileSiz  MemSiz   Flg Align
  LOAD           0x000000 0x0000000000000000 0x0000000000000000 0x000488 0x000488 R   0x1000
  LOAD           0x000488 0x0000000000001488 0x0000000000001488 0x000088 0x000088 R E 0x1000
  LOAD           0x000510 0x0000000000002510 0x0000000000002510 0x0001f0 0x0001f0 RW  0x1000
  LOAD           0x000700 0x0000000000003700 0x0000000000003700 0x000000 0x3d0900 RW  0x1000
```

**Comment by yschandra** [comment]
I am using upstream clang-9 and lld to compile Android-x86 (Android P based release) and see that I have the same problem with **/system/lib64/libdl.so**
**Error:**
```
[   11.472503] e2fsck: CANNOT LINK EXECUTABLE "/system/bin/e2fsck": invalid ELF file "/system/lib64/libdl.so" load segment[4]: p_offset (0x4000) + p_filesz (0x0) ( = 0x4000) past end of file (0x3880)
[   11.491438] e2fsck: linker: CANNOT LINK EXECUTABLE "/system/bin/e2fsck": invalid ELF file "/system/lib64/libdl.so" load segment[4]: p_offset (0x4000) + p_filesz (0x0) ( = 0x4000) past end of file (0x3880)
[   11.511039] e2fsck: e2fsck terminated by exit(1)
```
**libdl.so**
```
Program Headers:
  Type           Offset             VirtAddr           PhysAddr
                 FileSiz            MemSiz              Flags  Align
  PHDR           0x0000000000000040 0x0000000000000040 0x0000000000000040
                 0x0000000000000230 0x0000000000000230  R      0x8
  LOAD           0x0000000000000000 0x0000000000000000 0x0000000000000000
                 0x0000000000001034 0x0000000000001034  R      0x1000
  LOAD           0x0000000000002000 0x0000000000002000 0x0000000000002000
                 0x0000000000000330 0x0000000000000330  R E    0x1000
  LOAD           0x0000000000003000 0x0000000000003000 0x0000000000003000
                 0x00000000000001b8 0x00000000000001b8  RW     0x1000
  LOAD           0x0000000000004000 0x0000000000004000 0x0000000000004000
                 0x0000000000000000 0x0000000000001000  RW     0x1000
  DYNAMIC        0x0000000000003000 0x0000000000003000 0x0000000000003000
                 0x0000000000000110 0x0000000000000110  RW     0x8
  GNU_RELRO      0x0000000000003000 0x0000000000003000 0x0000000000003000
                 0x00000000000001b8 0x0000000000001000  R      0x1
  GNU_EH_FRAME   0x0000000000000d68 0x0000000000000d68 0x0000000000000d68
                 0x00000000000000b4 0x00000000000000b4  R      0x4
  GNU_STACK      0x0000000000000000 0x0000000000000000 0x0000000000000000
                 0x0000000000000000 0x0000000000000000  RW     0x0
  NOTE           0x0000000000000270 0x0000000000000270 0x0000000000000270
                 0x0000000000000020 0x0000000000000020  R      0x4

 Section to Segment mapping:
  Segment Sections...
   00
   01     .note.gnu.build-id .dynsym .gnu.version .gnu.version_d .gnu.hash .dynstr .rela.plt .eh_frame_hdr .eh_frame
   02     .text .plt
   03     .dynamic .got.plt
   04     .bss
   05     .dynamic
   06     .dynamic .got.plt
   07     .eh_frame_hdr
   08
   09     .note.gnu.build-id
```
When I use gold linker this problem goes away. The problem seems to stem from **.bss** being put into a separate loadable segment. As per spec **.bss** will not consume any file space, so **p_offset** in program header is kind of bogus which is set to **p_vaddr**, the linker trying validate with file size is not right in this case. I believe the generated elf is still valid in this case.

The gold linker seems to combine .bss with other section in to a single segment, which seem to pass the linker check. This is the reason the check passes and things work for me. 

Check in my version of linker [https://android.googlesource.com/platform/bionic/+/refs/tags/android-9.0.0_r52/linker/linker_phdr.cpp#648](url)

**Comment by yschandra** [comment]
For my problem tired a ugly work around adding linker flag to **libdl** '-Wl,--Tbss=0x270' hoping to move the **.bss** up next PHDR and it worked and linker check passes!
```
Program Headers:
  Type           Offset             VirtAddr           PhysAddr
                 FileSiz            MemSiz              Flags  Align
  PHDR           0x0000000000000040 0x0000000000000040 0x0000000000000040
                 0x0000000000000268 0x0000000000000268  R      0x8
  LOAD           0x0000000000000000 0x0000000000000000 0x0000000000000000
                 0x00000000000002a8 0x00000000000002a8  R      0x1000
  LOAD           0x0000000000001000 0x0000000000001000 0x0000000000001000
                 0x0000000000000000 0x0000000000001000  RW     0x1000
  LOAD           0x0000000000001000 0x0000000000002000 0x0000000000002000
                 0x0000000000000dc4 0x0000000000000dc4  R      0x1000
  LOAD           0x0000000000002000 0x0000000000003000 0x0000000000003000
                 0x0000000000000330 0x0000000000000330  R E    0x1000
  LOAD           0x0000000000003000 0x0000000000004000 0x0000000000004000
                 0x00000000000001b8 0x00000000000001b8  RW     0x1000
  DYNAMIC        0x0000000000003000 0x0000000000004000 0x0000000000004000
                 0x0000000000000110 0x0000000000000110  RW     0x8
  GNU_RELRO      0x0000000000003000 0x0000000000004000 0x0000000000004000
                 0x0000000000001000 0x0000000000001000  R      0x1
  GNU_EH_FRAME   0x0000000000001af8 0x0000000000002af8 0x0000000000002af8
                 0x00000000000000b4 0x00000000000000b4  R      0x4
  GNU_STACK      0x0000000000000000 0x0000000000000000 0x0000000000000000
                 0x0000000000000000 0x0000000000000000  RW     0x0
  NOTE           0x0000000000001000 0x0000000000002000 0x0000000000002000
                 0x0000000000000020 0x0000000000000020  R      0x4

 Section to Segment mapping:
  Segment Sections...
   00     
   01     
   02     .bss 
   03     .note.gnu.build-id .dynsym .gnu.version .gnu.version_d .gnu.hash .dynstr .rela.plt .eh_frame_hdr .eh_frame 
   04     .text .plt 
   05     .dynamic .got.plt 
   06     .dynamic 
   07     .dynamic .got.plt 
   08     .eh_frame_hdr 
   09     
   10     .note.gnu.build-id 
```

**Comment by yschandra** [comment]
More sane fix for my situation is to have a linker script put the .bss along with others in PT_LOAD segment
```
SECTIONS {
  . = SIZEOF_HEADERS;
  .bss : { *(.bss) }
  .dynamic : { *(.dynamic) }
  .got.plt : { *(.got.plt) }
}
```
Which did the trick and issue resolved. I believe android runtime linker needs to be fixed.
```
Program Headers:
  Type           Offset             VirtAddr           PhysAddr
                 FileSiz            MemSiz              Flags  Align
  PHDR           0x0000000000000040 0x0000000000000040 0x0000000000000040
                 0x00000000000001c0 0x00000000000001c0  R      0x8
  LOAD           0x0000000000000000 0x0000000000000000 0x0000000000000000
                 0x0000000000001300 0x0000000000001300  R E    0x1000
  LOAD           0x0000000000002000 0x0000000000002000 0x0000000000002000
                 0x00000000000011b8 0x00000000000011b8  RW     0x1000
  DYNAMIC        0x0000000000003000 0x0000000000003000 0x0000000000003000
                 0x0000000000000110 0x0000000000000110  RW     0x8
  NULL           0x0000000000000000 0x0000000000000000 0x0000000000000000
                 0x0000000000000000 0x0000000000000000         0x0
  GNU_EH_FRAME   0x0000000000000cf8 0x0000000000000cf8 0x0000000000000cf8
                 0x00000000000000b4 0x00000000000000b4  R      0x4
  GNU_STACK      0x0000000000000000 0x0000000000000000 0x0000000000000000
                 0x0000000000000000 0x0000000000000000  RW     0x0
  NOTE           0x0000000000000200 0x0000000000000200 0x0000000000000200
                 0x0000000000000020 0x0000000000000020  R      0x4

 Section to Segment mapping:
  Segment Sections...
   00     
   01     .note.gnu.build-id .dynsym .gnu.version .gnu.version_d .gnu.hash .dynstr .rela.plt .eh_frame_hdr .eh_frame .text .plt 
   02     .bss .dynamic .got.plt 
   03     .dynamic 
   04     
   05     .eh_frame_hdr 
   06     
   07     .note.gnu.build-id 
```

**Comment by DanAlbert** [comment]
@rprichard triaging to r22 to either confirm that it's fixed or still present. If it's still present we may want to check ToT, but that update would probably need to wait for r23 anyway.

**Comment by enh-google** [comment]
@DanAlbert / @rprichard / @pirama-arumuga-nainar --- have i misunderstood what the actual bug here is? is it actually strip rather than the linker that's the problem, so the "fix" is "switch to LLVM strip"?

**Comment by DanAlbert** [comment]
By my reading it's not confirmed whether that's a GNU strip bug or a bug elsewhere that only GNU strip is exposing. Does seem that using llvm-strip would solve the problem at least temporarily though.

**Comment by rprichard** [comment]
Upstream GNU strip accepted a change in Dec 2019 to avoid creating a p_offset that exceeded the file size, because doing so broke WSL's loader (and also breaks Bionic's loader). I got the impression that lld's output was unobjectionable but different enough from ld.bfd/gold to trigger this situation in GNU strip.

Apparently glibc is fine with a zero-file-size PT_LOAD p_offset that exceeds the file size. The p_offset must still be aligned, though, I guess (https://sourceware.org/bugzilla/show_bug.cgi?id=25237#c1)? Why not just ignore it entirely? 

AFAIK, we could fix Bionic to ignore p_offset if p_filesz is zero. Of course, that won't help with existing versions of Android, so we would still need a toolchain fix.

Switching to llvm-strip seems to avoid the problem.

**Comment by DanAlbert** [comment]
Okay, in that case, I think we should just close this as obsolete. GNU strip isn't gone from r22, but it's no longer used by any of our build systems and we've recommended that people migrate to llvm-strip. Both lld and llvm-strip are experimental in r21, and the changelog does recommend against using lld with GNU strip. Given that, I don't think there's justification for a backport here, and there's no need to update binutils in a future release because we've stopped using the broken tool.
